### PR TITLE
GeoJSON to binary arrays improvements

### LINKED ISSUE #702: Flat binary arrays followup
Resuming from #685, #690 

- **Index array length** - We should indeed make index arrays one element longer and include the length as final index.
- **Mixed coordinate dimensions**. If only some of the coordinates have 3 dimensions, some values in `positions` will be off by 1. If it's important to support varying coordinate dimensions, this needs to be fixed.
- **Feature properties**. The options are 1) convert numeric properties to binary arrays, 2) return an array of property objects with each geometry type 3) not returning properties. It might be worthwhile to implement 1 and 2, and let usage be user-decided.

	For 2, should `objectIds` point to the index in its local `properties` array? I.e. here, `points.objectIds` points to `points.properties`?
	
	```js
	{
	  points: {
	    // Array of x, y or x, y, z positions
	    positions: {value: Float32Array, size: coordLength},
    	// Array of original feature indexes by vertex
	    objectIds: {value: Uint32Array, size: 1},
		// Array of property objects
		properties: [
			{...}
		]
	  },
	  ...
	}
	```

### PR DESCRIPTION
**Pull request by kylebarron**
@ibgreen 

- [x] Index array length fixed, where the last value is coordinate length
- [x] Mixed coordinate dimensions: Should be fixed; if mixed coordinate lengths, `0` is left for the missing values
- [x] Numeric properties added to typed arrays, where each property is contained within the `points/lines/polygons.numericProps` object. It was helpful to make it its own object, since I loop over the keys of `numericProps` a couple times.
- [x] Calculations from the first pass are combined with user-provided options. So, for example, if a user didn't want any numeric properties in binary arrays, they could pass `numericProps = []`.
- [x] Feature properties (minus any numeric keys) are added to `properties` arrays by geometry type. If the user passes `numericProps = []`, all properties are kept as JSON.
- [x] Both `featureIndex` and `globalFeatureIndex` arrays are created.
- [x] Add new tests/fix changes to existing tests
- [x] Use smallest data type necessary for TypedArrays
- [ ] An option for user-defined missing values for `positions` and numeric feature properties? I.e. if non-zero, fill this as the default value for `positions` when it's initialized, and then I think the third coordinate won't get overwritten when filling a 2D point.

Closes #702

### CODE REVIEW

**Comment by kylebarron** [comment]
@ibgreen any feedback?

File: modules/gis/src/lib/geojson-to-binary.js [hunk 1]
```diff
@@ -1,13 +1,13 @@
 // Convert GeoJSON features to flat binary arrays
 export function geojsonToBinary(features, options = {}) {
   const firstPassData = firstPass(features);
-  return secondPass(features, firstPassData, options);
+  return secondPass(features, {...firstPassData, ...options});
```
Comment (ibgreen) [inline modules/gis/src/lib/geojson-to-binary.js]: Nit: This is a bit too compact for my taste, as these are very different objects with just a small overlap of fields. I would more explicitly copy any auto deduced options into the options object:

```
return secondPass(features, firstPassData, {maxCoordLength: firstPassData.maxCoordLength, ..., ...options);
```

Being more strict with "types" will also help as we are gradually introducing typescript type definition files in this repo.

File: modules/gis/src/lib/geojson-to-binary.js [hunk 2]
```diff
@@ -16,31 +16,28 @@ function firstPass(features) {
   let polygonPositions = 0;
   let polygonObjects = 0;
   let polygonRings = 0;
-  let maxCoordLength = 2;
+  const coordLengths = new Set();
```
Comment (ibgreen) [inline modules/gis/src/lib/geojson-to-binary.js]: Interesting approach.

File: modules/gis/src/lib/geojson-to-binary.js [hunk 3]
```diff
@@ -73,87 +68,133 @@ function firstPass(features) {
...
         }
         break;
       default:
         throw new Error(`Unsupported geometry type: ${geometry.type}`);
     }
+
+    if (feature.properties) {
+      for (const key in feature.properties) {
+        const val = feature.properties[key];
+        numericProps[key] = numericProps[key] ? isNumeric(val) : numericProps[key];
```
Comment (ibgreen) [inline modules/gis/src/lib/geojson-to-binary.js]: The logic may well be correct, but I get confused here. Do we want this to be true set only if all the values are numeric? Maybe the following is easier to follow?

`
// initalize
for (const key in feature.properties) {propIsNumericKey[key] = true;}
...
// each iteration, check if still considered numeric
if (propIsNumericKey[key]) {
  propIsNumericKey[key] = Number.isFinite(val);
}

File: modules/gis/src/lib/geojson-to-binary.js [hunk 4]
```diff
@@ -73,87 +68,133 @@ function firstPass(features) {
...
+    numericProps: Object.keys(numericProps).filter(k => numericProps[k])
   };
 }
 
 // Second scan over GeoJSON features
 // Fills coordinates into pre-allocated typed arrays
-function secondPass(features, firstPassData, options = {}) {
+function secondPass(features, options = {}) {
   const {
     pointPositions,
```
Comment (ibgreen) [inline modules/gis/src/lib/geojson-to-binary.js]: Idea: should we call these `pointPositionCount` etc? As I read this I keep asking myself if this is an array of positions or just the count.

File: modules/gis/src/lib/geojson-to-binary.js [hunk 5]
```diff
@@ -73,87 +68,133 @@ function firstPass(features) {
...
-  } = firstPassData;
-  const {PositionDataType = Float32Array} = options;
+    polygonRings,
+    numericProps,
+    PositionDataType = Float32Array
+  } = options;
   const points = {
     positions: new PositionDataType(pointPositions * coordLength),
-    objectIds: new Uint32Array(pointPositions)
+    globalFeatureIndex: new Uint32Array(pointPositions),
```
Comment (ibgreen) [inline modules/gis/src/lib/geojson-to-binary.js]: I know we decided to do this later, but if we could add a count of the features:

`globalFeatureIndex: globalFeatureCount > 65535 ? new Uint32Array(pointPositions) : new Uint16Array(pointPositions),`

File: modules/gis/src/lib/geojson-to-binary.js [hunk 6]
```diff
@@ -73,87 +68,133 @@ function firstPass(features) {
...
+    properties: []
   };
   const polygons = {
-    polygonIndices: new Uint32Array(polygonObjects),
-    primitivePolygonIndices: new Uint32Array(polygonRings),
+    polygonIndices: new Uint32Array(polygonObjects + 1),
+    primitivePolygonIndices: new Uint32Array(polygonRings + 1),
     positions: new PositionDataType(polygonPositions * coordLength),
-    objectIds: new Uint32Array(polygonPositions)
+    globalFeatureIndex: new Uint32Array(polygonPositions),
```
Comment (ibgreen) [inline modules/gis/src/lib/geojson-to-binary.js]: For consistency as other arrays are named with plurals, these would be either `...Indices` or `...Ids`?

**Comment by ibgreen** [comment]
@kylebarron Just mark the request as "ready for review" and ping me and I will land.

File: modules/gis/src/lib/geojson-to-binary.js (hunk 2 above)
Comment (kylebarron) [reply modules/gis/src/lib/geojson-to-binary.js]: I changed it because I thought it might be helpful to know not only that the max coordinate length is 3, but also that there are mixed coordinates. Having the `Set` would tell me both pieces of information. That said, right now I don't think I use that information. 

Do you think there's a performance hit of using a `Set`?

Comment (ibgreen) [reply modules/gis/src/lib/geojson-to-binary.js]: > Do you think there's a performance hit of using a Set?

Nah, its fine.

File: modules/gis/src/lib/geojson-to-binary.js (hunk 5 above)
Comment (kylebarron) [reply modules/gis/src/lib/geojson-to-binary.js]: I updated this to use `Uint16Array` where possible both for the feature indices and for the `pathIndices`, `polygonIndices`, `primitivePolygonIndices`. (Aside, now that I think about it more, `polygonRingIndices` seems like a more apt name, since the term "linear ring" is formalized in the GeoJSON spec).

I also updated to count features by geometry type, so if there are 66,000 features, but split equally as 22,000 of point/line/polygon, the `globalFeatureIndex` arrays would be `Uint32`, but the local ones would be `Uint16`

Comment (ibgreen) [reply modules/gis/src/lib/geojson-to-binary.js]: Awesome! Just consider what test coverage you think we need for this, and since you are attending to a lot of little details, I'd recommend making sure documentation is updated as you go  through each of them 

(This is a general comment, I haven't checked if / what you added in tests/docs in the last commits).

File: modules/gis/src/lib/geojson-to-binary.js (hunk 3 above)
Comment (kylebarron) [reply modules/gis/src/lib/geojson-to-binary.js]: > Do we want this to be true only if all the values in all features are numeric? 

I think so, yes.

> If so maybe the following is easier to follow?

That snippet is initializing based on the properties in the first feature? The problem with that is that not all features are required to have all keys. With MVT in particular, if I'm not mistaken, a property could be defined in only a single feature within a layer.

So if we want `numericKey` to be `true` if the values for that prop in all features are numeric **and** exist, then initializing based on the properties in the first feature are fine. Otherwise, to initialize based on all properties I think it would be necessary to loop over all features just to do the initialization.

I'm currently interpreting `numericKey` should be `true` if the values for that prop in all features are either numeric **or** do not exist. So I need to keep track of 1) have I seen this key in a previous feature, 2) has it been non-numeric in any previous feature.

It might be good to add a comment like
```js
// If property has been numeric in all previous features in which the
// property existed, check if numeric in this feature
```

Comment (ibgreen) [reply modules/gis/src/lib/geojson-to-binary.js]: Sounds good. This can get pretty sophisticated, so you should draw the line here for now and then we can come back and make further improvements if we feel so inclined.

For later:
- An approach used in binary columnar tables is to optionally generate a second `null` array to track which objects are not present (e.g. a `Uint8Array`).
- We could detect the type of the array:
   -  For each element check if integer. If row is all integer, generate `Uint32Array`
   - Track max/min values of all values
    - if (-32000, 32000), generate `Uint16Array`
    - If (-128, 127) generate `Uint8Array`
    - etc

Maybe create a follow up tasks with such ideas for a later date?

Comment (kylebarron) [reply modules/gis/src/lib/geojson-to-binary.js]: > This can get pretty sophisticated, so you should draw the line here for now and then we can come back and make further improvements if we feel so inclined.

Good idea, it's clear this can be a black hole.

One more idea I thought of that can be a future task: should `numericKeys` depend on the geometry type?

For example, if I have a set of GeoJSON features where a numeric property `x` only exists on `Point` features, and I have 10 `Point` features but 1000 `Polygon` features, the current code instantiates a `Float32Array` for _every Polygon vertex_, when every vertex is missing data. 

I don't think this would necessarily be uncommon either; I have a `mountain_peak` layer of points in my MVTs with a numeric `elevation` value, but no other layer has that property, so there would be `# line positions + # polygon positions` extra empty data just for that property.

Comment (ibgreen) [reply modules/gis/src/lib/geojson-to-binary.js]: Yes that is a good observation, that can be another optimization.

Right now I think we should be concerned with the "combinatorial explosion" of options + input dependent outputs, specifically relating to the testability and ease-of-use of it all. 

Less optionality is probably good for now

File: modules/gis/src/lib/geojson-to-binary.js (hunk 1 above)
Comment (kylebarron) [reply modules/gis/src/lib/geojson-to-binary.js]: What do you think about that revision? It's more explicit and essentially defines three options that a user can provide:

- `coordLength` force a specific coordinate length.
- `numericPropKeys` could pass `[]` to prevent creating any numeric property arrays
- `PositionDataType` could pass `Float64Array` to store positions at higher precision

Comment (ibgreen) [reply modules/gis/src/lib/geojson-to-binary.js]: +1

File: modules/gis/src/lib/geojson-to-binary.js (hunk 3 above)
Comment (kylebarron) [reply modules/gis/src/lib/geojson-to-binary.js]: Yes, you're right. It's easy to go on a feature-adding spree. If there's anything you think should be taken out let me know.

Now I'm pivoting to adding more tests... My initial approach would be to add tests to `firstPass` and `secondPass` separately, to more easily test edge cases. Do you know if it's possible to write tests for a non-`exported` function in JS? Or is it ok to export the functions for use in tests? I saw one approach on the web like
```js
export const _testing = [firstPass, secondPass]
```

Comment (ibgreen) [reply modules/gis/src/lib/geojson-to-binary.js]: Yes I sometimes add a `TEST_EXPORTS`: `const TEST_EXPORTS = {firstPass, secondPass}`

Also by using an object you can then do:
```
import {TEST_EXPORTS} from ...; 
const {firstPass, secondPass} = TEST_EXPORTS;
```

**Comment by kylebarron** [comment]
@ibgreen 
I think I covered a decent number of edge cases in the tests I added. I also updated the docs. Do you think they're clear enough?

File: modules/gis/docs/api-reference/geojson-to-binary.md [hunk 7]
```diff
@@ -19,51 +19,92 @@ const geoJSONfeatures = await load(url, MVTLoader, loaderOptions);
...
 ```
 
 ## Outputs
 
-### TypedArrays
-
 `geojsonToBinary` returns an object containing typed arrays sorted by geometry
-type. `positions` corresponds to 2D or 3D coordinates; `objectIds` returns the
-index of the vertex in the initial `features` array.
+type. `positions` is a flat array of coordinates; `globalFeatureIds` references
```
Comment (ibgreen) [inline modules/gis/docs/api-reference/geojson-to-binary.md]: For easier reading and a cleaner look, use a markdown table, or at least a bulleted list?

File: modules/gis/docs/api-reference/geojson-to-binary.md [hunk 8]
```diff
@@ -19,51 +19,92 @@ const geoJSONfeatures = await load(url, MVTLoader, loaderOptions);
...
+    }
+    // Array of objects with non-numeric properties from Point geometries
+    properties: [{PointFeatureProperties}]
   },
   lines: {
-    // Indices within positions of the start of each individual LineString
-    pathIndices: {value: Uint32Array, size: 1},
     // Array of x, y or x, y, z positions
-    positions: {value: Float32Array, size: coordLength},
+    positions: {value: PositionDataType, size: coordLength},
```
Comment (ibgreen) [inline modules/gis/docs/api-reference/geojson-to-binary.md]: We are in the process of introducing typescript defintions for loaders.gl 2.2. With minimal extra effort, we could actually turn what you wrote here into typescript types in a `geojson-to-binary.d.ts`. Look for other `.d.ts` files to see how it is done.

File: modules/gis/docs/api-reference/geojson-to-binary.md [hunk 9]
```diff
@@ -19,51 +19,92 @@ const geoJSONfeatures = await load(url, MVTLoader, loaderOptions);
...
 ```
 
 ## Options
 
-| Option           | Type                             | Default        | Description                         |
-| ---------------- | -------------------------------- | -------------- | ----------------------------------- |
-| PositionDataType | `Float32Array` or `Float64Array` | `Float32Array` | Data type used for positions arrays |
+| Option           | Type     | Default           | Description                                                                                                                                             |
+| ---------------- | -------- | ----------------- | ------------------------------------------------------------------------------------------------------------------------------------------------------- |
+| PositionDataType | `class`  | `Float32Array`    | Data type used for positions arrays.                                                                                                                    |
```
Comment (ibgreen) [inline modules/gis/docs/api-reference/geojson-to-binary.md]: This reminds me: It is best if the option values are representable in JSON, i.e. numbers, strings, etc. So you could support `"float32"` and `"float64"`, or `positionPrecision: 32 | 64`

The reason is that loaders.gl is integrated with deck.gl, and deck.gl exposes loader options in its API, and for instance pydeck and non-JS language bindings go through the deck.gl JSON API intermediary.

File: modules/gis/src/lib/geojson-to-binary.js [hunk 10]
```diff
@@ -1,159 +1,251 @@
...
   const firstPassData = firstPass(features);
-  return secondPass(features, firstPassData, options);
+  return secondPass(features, firstPassData, {
+    coordLength: options.coordLength || firstPassData.coordLength,
+    numericPropKeys: options.numericPropKeys || firstPassData.numericPropKeys,
+    PositionDataType: options.PositionDataType || Float32Array
+  });
 }
 
+export const TEST_EXPORTS = {
```
Comment (ibgreen) [inline modules/gis/src/lib/geojson-to-binary.js]: Nit: I normally put this last in file

File: modules/gis/test/geojson-to-binary.spec.js [hunk 11]
```diff
@@ -1,110 +1,310 @@
...
+  const {points, lines, polygons} = secondPass(features, firstPassData, options);
+
+  // 3D size
+  t.equal(points.positions.size, 3);
+  t.equal(lines.positions.size, 3);
+  t.equal(polygons.positions.size, 3);
+
+  // Test value equality, missing third dimension imputed as 0
+  t.deepEqual(points.positions.value, [100, 0, 1, 100, 0, 0, 101, 1, 0]);
+  t.deepEqual(lines.positions.value, [
```
Comment (ibgreen) [inline modules/gis/test/geojson-to-binary.spec.js]: Use `// prettier-ignore` just before line 125 to prevent auto formatting?
